xarray/core/variable.py

```diff
@@ -412,15 +473,15 @@ def __setitem__(self, key, value):
 
         See __getitem__ for more details.
         """
-        key = self._item_key_to_tuple(key)
+        dims, index_tuple = self._broadcast_indexes(key)
         if isinstance(self._data, dask_array_type):
             raise TypeError("this variable's data is stored in a dask array, "
                             'which does not support item assignment. To '
                             'assign to this variable, you must first load it '
                             'into memory explicitly using the .load_data() '
                             'method or accessing its .values attribute.')
-        data = orthogonally_indexable(self._data)
-        data[key] = value
+        data = broadcasted_indexable(self._data)
+        data[index_tuple] = value
```

Comment:
`value` should be broadcast against the indexing `key` if possible.

If it's an `xarray.Variable`, we should call `value.set_dims(dims)` on it. If it's a NumPy/Dask array, I think we can safely ignore it and let NumPy/Dask handle broadcasting.